File: src/dotnet/Src/ButterflyNet.Parser/Satisfiers/CurrentScopeMustMatchOrBeParagraphSatisfier.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ButterflyNet.Parser.Satisfiers {
	public class CurrentScopeMustMatchOrBeParagraphSatisfier : ISatisfier {
		private readonly IEnumerable<Type> scopeTypes;

		public CurrentScopeMustMatchOrBeParagraphSatisfier(params Type[] scopeTypes) {
			this.scopeTypes = scopeTypes.Concat(new[] { ScopeTypeCache.Paragraph });
		}

		public bool IsSatisfiedBy(ParseContext context) {
			var currentScope = context.Scopes.PeekOrDefault();
			return currentScope != null && scopeTypes.Contains(currentScope.GetType());
		}
	}
}

File: src/js/src/satisfiers/CurrentScopeMustMatchOrBeParagraphSatisfier.js
function CurrentScopeMustMatchOrBeParagraphSatisfier() {
	var scopeTypes = toArray(arguments);
	scopeTypes.push(ScopeTypeCache.paragraph);
	
	this.isSatisfiedBy = function(context) {
		return !context.scopes.isEmpty() && contains(scopeTypes, context.scopes.peek().type);
	};
}

extend(Satisfier, CurrentScopeMustMatchOrBeParagraphSatisfier);